python/selfie-lib/selfie_lib/ArrayMap.py

```diff
@@ -0,0 +1,148 @@
+from collections.abc import Set, Sequence, Iterator, Mapping
+from typing import List, TypeVar, Union, overload, Sequence as TypeSequence
+from functools import cmp_to_key
+
+T = TypeVar('T')
+V = TypeVar('V')
+K = TypeVar('K', bound='Comparable')
+
+class Comparable:
+    def __lt__(self, other: 'Comparable') -> bool: ...
+    def __le__(self, other: 'Comparable') -> bool: ...
+    def __gt__(self, other: 'Comparable') -> bool: ...
+    def __ge__(self, other: 'Comparable') -> bool: ...
+
+class ListBackedSet(Set[T], Sequence[T]):
+    def __init__(self):
+        self._list: List[T] = []
+
+    @overload
+    def __getitem__(self, index: int) -> T: ...
+    
+    @overload
+    def __getitem__(self, index: slice) -> TypeSequence[T]: ...
+
+    def __getitem__(self, index: Union[int, slice]) -> Union[T, TypeSequence[T]]:
+        if isinstance(index, int):
+            return self._list[index]
+        elif isinstance(index, slice):
+            return self._list[index]
+        else:
+            raise TypeError("Index must be int or slice")
+
+    def __len__(self) -> int:
+        return len(self._list)
+
+    def __iter__(self) -> Iterator[T]:
+        return iter(self._list)
+
+    def __contains__(self, item: object) -> bool:
+        return item in self._list
+
+class ArraySet(ListBackedSet[K]):
+    def __init__(self, data: List[K]):
+        self.data = data
+        self.sort_data()
+
+    def sort_data(self):
+        if self.data and isinstance(self.data[0], str):
+            self.data.sort(key=cmp_to_key(self.string_slash_first_comparator))
+        else:
+            self.data.sort()
+
+    def string_slash_first_comparator(self, a, b):
+        if a == '/':
+            return -1
+        elif b == '/':
+            return 1
+        else:
+            return (a > b) - (a < b)
+
+    def __len__(self):
+        return len(self.data)
+
+    @overload
+    def __getitem__(self, index: int) -> K: ...
+
+    @overload
+    def __getitem__(self, index: slice) -> 'ArraySet[K]': ...
+
+    def __getitem__(self, index: Union[int, slice]) -> Union[K, 'ArraySet[K]']:
+        if isinstance(index, int):
+            return self.data[index]
+        elif isinstance(index, slice):
+            sliced_data = self.data[index]
+            return ArraySet(sliced_data)
+        else:
+            raise TypeError("Index must be int or slice")
+
+    def __contains__(self, item: object) -> bool:
+        if not isinstance(item, type(self.data[0])):
+            return False
+        return super().__contains__(item)
```

Comment:
If you make the changes to `ListBackedSet` mentioned above, then none of these should need to be implemented here.